File: 268/missing-number.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var missingNumber = function (nums) {
    let hash = [];
    let len = nums.length;
    for (let i = 0; i < len; i++) {
        hash[nums[i]] = true;
    }

    for (let j = 0; ; j++) {
        if (!hash[j])
            return j;
    }
};

var missingNumber = function (nums) {
    let sum = 0;
    for (let i = 0; i <= nums.length; i++) {
        sum ^= i;
    }
    return nums.reduce((x, y) => x ^ y, sum);
};

File: 268/MissingNumber.scala
object MissingNumber {
  def missingNumber(nums: Array[Int]): Int = {
    val t = (0 to nums.length).foldLeft(0)(_ ^ _)
    nums.foldLeft(t)(_ ^ _)
  }

  def main(args: Array[String]): Unit = {
    val nums = Array(9,6,4,2,3,5,7,0,1)
    println(missingNumber(nums))
  }
}
